README.rst
shield-python is a set of Python libraries that allows easy access to the
shield peer-to-peer cryptocurrency client API.

Documentation
===========================

Documentation can be found here, or in the source archive. It is built
using Sphinx:

http://laanwj.github.com/shield-python/doc/

Installation instructions
===========================

shield-python uses setuptools for the install script. There are no dependencies apart from Python itself.

::

  $ python setup.py build
  $ python setup.py install

TODO
======
These things still have to be added:

- SSL support (including certificate verification) for managing remote shield daemons.



Changelog
2013-03-30  Wladimir van der Laan <laanwj@gmail.com>

	* Release 0.3

2013-03-30 16:13:06 +0100 Aviad Reich 

	* add keypoolrefill method (HEAD, origin/master, master)

2013-03-20 00:33:59 +0200 Aviad Reich 

	* raw transaction support added (rawtransactions)

2013-03-21 14:41:33 +0200 Aviad Reich 

	* cherry pick to fix dont_raise so it only blocks the correct exceptions to return a boolean (origin/wallet_locking)

2013-03-29 10:15:09 +0300 Aviad Reich 

	* ignore dist and python_egg

2013-03-23 19:08:46 +0000 Aviad Reich 

	* deprecated getblocknumber - use getblockcount instead

2013-03-19 23:01:18 +0200 Aviad Reich 

	* add wallet locking/unlocking support

2013-03-19 14:42:26 +0200 Aviad Reich 

	* in listtransactions, iterate all accounts if account is None (origin/listtransactions)

2013-03-19 11:16:38 +0200 Aviad Reich 

	* listaccounts can return a dictionary from account names to their balance if the param as_dict is True, update tests. (origin/listaccounts)

2013-03-19 09:44:20 +0200 Aviad Reich 

	* add tests for listunspent and getmininginfo

2013-03-18 08:18:23 +0200 Aviad Reich 

	* refactor __repr__

2013-03-17 09:43:44 +0200 Aviad Reich 

	* flake8 style conventions fixes

2013-03-18 00:13:42 +0200 Aviad Reich 

	* add listunspent method

2013-03-17 23:46:44 +0200 Aviad Reich 

	* add support for getmininginfo

2013-03-17 16:37:09 +0200 Aviad Reich 

	* Issue #15: add from parameter to listtransactions

2013-03-15 22:14:10 +0100 David Barton 

	* Fixed usage.rst broken links

2013-03-14 09:26:23 +0200 Aviad Reich 

	* connect_to_local(): use default testnet port, when the config file has testnet=x

2013-02-10 17:35:14 +0100 Adrian 

	* Add getrawtransaction method

2013-01-16 03:28:11 +0100 Thomas Steen Rasmussen 

	* fix bug when rpc reply error is null in json

2012-12-12 00:14:34 +0100 Benoît HERVIER 

	* There isn't always an error key in resp dict

2012-12-12 00:10:41 +0100 Benoît HERVIER 

	* Allow use of https

2012-10-29 16:46:18 +0100 Harald Schilly 

	* assert that we are really in the testnet

2012-10-28 23:25:25 +0100 Harald Schilly 

	* optionally allow to pass in the path to the shield.conf file if it is at a non-standard path

2012-10-28 21:52:48 +0100 Harald Schilly 

	* fixing tests

2012-10-23 02:57:40 +0200 Thomas Steen Rasmussen 

	* Add support for the 'verifymessage' RPC command

2012-10-05 18:45:50 +0200 Wladimir J. van der Laan 

	* Switch to jgarzik's jsonrpc implementation

2012-10-05 18:32:23 +0200 Wladimir J. van der Laan 

	* Use RPC error codes from shield core source

2012-07-15 12:07:53 +0400 Alexander Petrovich 

	* Added listsinceblock, getblock, getblockhash

2011-09-04 15:17:44 +0200 Jure Vrscaj 

	* Added sendmany command.

2011-08-24 14:53:53 +0200 Wladimir J. van der Laan 

	* allow listtransactions() (default to '' account), add test for gettransaction

2012-10-23  Thomas Steen Rasmussen <thomas@gibfest.dk>

	* Add "verifymessage" command

2011-08-24  Wladimir van der Laan <laanwj@gmail.com>

	* Add metadata for PyPi packaging

	* Add Changelog

	* Rename package to 'shieldrpc' instead of 'shield', makes more sense as this is a RPC
	  client not a shield client implementation.

	* Add COPYING (MIT license)

	* Add "gettransaction" command (thanks chopachom)


src/shieldrpc/exceptions.py
# Copyright (c) 2010 Witchspace <witchspace81@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
"""
Exception definitions.
"""


class shieldException(Exception):
    """
    Base class for exceptions received from shield server.

    - *code* -- Error code from ``shieldd``.
    """
    # Standard JSON-RPC 2.0 errors
    INVALID_REQUEST  = -32600,
    METHOD_NOT_FOUND = -32601,
    INVALID_PARAMS   = -32602,
    INTERNAL_ERROR   = -32603,
    PARSE_ERROR      = -32700,

    # General application defined errors
    MISC_ERROR                  = -1  # std::exception thrown in command handling
    FORBIDDEN_BY_SAFE_MODE      = -2  # Server is in safe mode, and command is not allowed in safe mode
    TYPE_ERROR                  = -3  # Unexpected type was passed as parameter
    INVALID_ADDRESS_OR_KEY      = -5  # Invalid address or key
    OUT_OF_MEMORY               = -7  # Ran out of memory during operation
    INVALID_PARAMETER           = -8  # Invalid, missing or duplicate parameter
    DATABASE_ERROR              = -20 # Database error
    DESERIALIZATION_ERROR       = -22 # Error parsing or validating structure in raw format

    # P2P client errors
    CLIENT_NOT_CONNECTED        = -9  # shield is not connected
    CLIENT_IN_INITIAL_DOWNLOAD  = -10 # Still downloading initial blocks

    # Wallet errors
    WALLET_ERROR                = -4  # Unspecified problem with wallet (key not found etc.)
    WALLET_INSUFFICIENT_FUNDS   = -6  # Not enough funds in wallet or account
    WALLET_INVALID_ACCOUNT_NAME = -11 # Invalid account name
    WALLET_KEYPOOL_RAN_OUT      = -12 # Keypool ran out, call keypoolrefill first
    WALLET_UNLOCK_NEEDED        = -13 # Enter the wallet passphrase with walletpassphrase first
    WALLET_PASSPHRASE_INCORRECT = -14 # The wallet passphrase entered was incorrect
    WALLET_WRONG_ENC_STATE      = -15 # Command given in wrong wallet encryption state (encrypting an encrypted wallet etc.)
    WALLET_ENCRYPTION_FAILED    = -16 # Failed to encrypt the wallet
    WALLET_ALREADY_UNLOCKED     = -17 # Wallet is already unlocked

    def __init__(self, error):
        Exception.__init__(self, error['message'])
        self.code = error['code']


class TransportException(Exception):
    """
    Class to define transport-level failures.
    """
    def __init__(self, msg, code=None, protocol=None, raw_detail=None):
        self.msg = msg
        self.code = code
        self.protocol = protocol
        self.raw_detail = raw_detail
        self.s = """
        Transport-level failure: {msg}
        Code: {code}
        Protocol: {protocol}
        """.format(msg=msg, code=code, protocol=protocol)

    def __str__(self):
        return self.s


##### General application defined errors
class SafeMode(shieldException):
    """
    Operation denied in safe mode (run ``shieldd`` with ``-disablesafemode``).
    """


class JSONTypeError(shieldException):
    """
    Unexpected type was passed as parameter
    """
InvalidAmount = JSONTypeError  # Backwards compatibility


class InvalidAddressOrKey(shieldException):
    """
    Invalid address or key.
    """
InvalidTransactionID = InvalidAddressOrKey  # Backwards compatibility


class OutOfMemory(shieldException):
    """
    Out of memory during operation.
    """


class InvalidParameter(shieldException):
    """
    Invalid parameter provided to RPC call.
    """


##### Client errors
class ClientException(shieldException):
    """
    P2P network error.
    This exception is never raised but functions as a superclass
    for other P2P client exceptions.
    """


class NotConnected(ClientException):
    """
    Not connected to any peers.
    """


class DownloadingBlocks(ClientException):
    """
    Client is still downloading blocks.
    """


##### Wallet errors
class WalletError(shieldException):
    """
    Unspecified problem with wallet (key not found etc.)
    """
SendError = WalletError  # Backwards compatibility


class InsufficientFunds(WalletError):
    """
    Insufficient funds to complete transaction in wallet or account
    """


class InvalidAccountName(WalletError):
    """
    Invalid account name
    """


class KeypoolRanOut(WalletError):
    """
    Keypool ran out, call keypoolrefill first
    """


class WalletUnlockNeeded(WalletError):
    """
    Enter the wallet passphrase with walletpassphrase first
    """


class WalletPassphraseIncorrect(WalletError):
    """
    The wallet passphrase entered was incorrect
    """


class WalletWrongEncState(WalletError):
    """
    Command given in wrong wallet encryption state (encrypting an encrypted wallet etc.)
    """


class WalletEncryptionFailed(WalletError):
    """
    Failed to encrypt the wallet
    """


class WalletAlreadyUnlocked(WalletError):
    """
    Wallet is already unlocked
    """


# For convenience, we define more specific exception classes
# for the more common errors.
_exception_map = {
    shieldException.FORBIDDEN_BY_SAFE_MODE: SafeMode,
    shieldException.TYPE_ERROR: JSONTypeError,
    shieldException.WALLET_ERROR: WalletError,
    shieldException.INVALID_ADDRESS_OR_KEY: InvalidAddressOrKey,
    shieldException.WALLET_INSUFFICIENT_FUNDS: InsufficientFunds,
    shieldException.OUT_OF_MEMORY: OutOfMemory,
    shieldException.INVALID_PARAMETER: InvalidParameter,
    shieldException.CLIENT_NOT_CONNECTED: NotConnected,
    shieldException.CLIENT_IN_INITIAL_DOWNLOAD: DownloadingBlocks,
    shieldException.WALLET_INSUFFICIENT_FUNDS: InsufficientFunds,
    shieldException.WALLET_INVALID_ACCOUNT_NAME: InvalidAccountName,
    shieldException.WALLET_KEYPOOL_RAN_OUT: KeypoolRanOut,
    shieldException.WALLET_UNLOCK_NEEDED: WalletUnlockNeeded,
    shieldException.WALLET_PASSPHRASE_INCORRECT: WalletPassphraseIncorrect,
    shieldException.WALLET_WRONG_ENC_STATE: WalletWrongEncState,
    shieldException.WALLET_ENCRYPTION_FAILED: WalletEncryptionFailed,
    shieldException.WALLET_ALREADY_UNLOCKED: WalletAlreadyUnlocked,
}


def wrap_exception(error):
    """
    Convert a JSON error object to a more specific shield exception.
    """
    # work around to temporarily fix https://github.com/shield/shield/issues/3007
    if error['code'] == shieldException.WALLET_ERROR and error['message'] == 'Insufficient funds':
        error['code'] = shieldException.WALLET_INSUFFICIENT_FUNDS
    return _exception_map.get(error['code'], shieldException)(error)
